eos-live-boot-overlayfs-setup

```diff
@@ -47,18 +44,41 @@ setup_ostree_flatpak_overlay() {
     # via the /var/lib/flatpak symlink.
 }
 
-# Everything but /sysroot/{ostree,flatpak} is pretty straightforward:
-overlay_dirs="bin boot endless home lib opt ostree root sbin srv sysroot/home var"
-for dir in $overlay_dirs $EXTRA_PATHS; do
-    [ -d /$dir ] || continue
-    [ -d /run/eos-live/$dir ] && continue
+setup_overlay() {
+    local dir=$1
+    [ -d /$dir ] || return
+    [ -d /run/eos-live/$dir ] && return
     # If the directory is a symlink, assume it's pointing to a location
     # covered by another top level overlay
-    [ -L /$dir ] && continue
+    [ -L /$dir ] && return
     mkdir -p /run/eos-live/$dir /run/eos-live/$dir-workdir
     mount -t overlay -o \
         rw,upperdir=/run/eos-live/$dir,lowerdir=/$dir,workdir=/run/eos-live/$dir-workdir \
         eos-live-$dir /$dir
+}
+
+setup_overlay etc
+
+# If we booted and systemd was unable to find or create a machine-id,
+# it will create one at /run/machine-id and bind mount it at /etc/machine-id.
+# But now that we just setup an overlay at /etc, we lost that bind mount, so
+# we're back to having no machine-id.
+#
+# In addition to recovering the machine-id, we also want to use this
+# opportunity to store it in the overlayfs, in case the overlayfs is backed
+# by persistent storage.
+#
+# Both goals can be satisfied in simplistic fashion by copying systemd's
+# boot-generated machine-id into place, in the case where we didn't find one
+# in the overlay.
+if [[ ! -s /etc/machine-id && -s /run/machine-id ]]; then
+  cp -v /run/machine-id /etc/machine-id
```

Comment:
In this implementation there is a window of time where `/etc/machine-id` becomes empty once again. It's probably not a big problem but we could imagine doing this the other way round:

* Before setting up the overlay, if `/etc/machine-id` is a symlink to `/run/machine-id`, then copy `/run/machine-id` to `/run/eos-live/etc/machine-id`
* Now set up the overlay

Worthwhile? I'm not sure.

Reply:
> Before setting up the overlay, if /etc/machine-id is a symlink to /run/machine-id

It always will be as we're booting live media here.

There are other ways of detecting and avoiding the gap though, and I just pushed the simplest approach I can think of. Please take a look

Reply:
Good point.

I'd never seen `mount --move` before – nice solution.